src/Dscp/Educator/Validate.hs

```diff
@@ -2,41 +2,104 @@
+        , svfSubmissionSigKey :: !PublicKey
+        , svfSubmissionSig    :: !SubmissionSig }
+    deriving (Eq)
+
+instance Show BlockValidationFailure where
+    show = toString . pretty
+
+instance Buildable BlockValidationFailure where
+    build (MerkleSignatureMismatch {..}) =
+      "Merkle tree root signature mismatch. Expected " `mappend` show bvfExpectedSig
+      `mappend` " got " `mappend` show bvfActualMerkleSig
+    build (SubmissionInvalid x) = build x
+
+instance Buildable SubmissionValidationFailure where
+    build (SubmissionPublicKeyMismatch {..}) =
+      "Submission public key address mismatch. Expected " `mappend` show svfExpectedPubKey `mappend`
+      " got " `mappend` show svfActualPubKey
+    build (SubmissionSignatureMismatch {..}) =
+      "Submission signature mismatch. Submission data " `mappend` show svfSubmissionHash
+      `mappend` " with signature " `mappend` show svfSubmissionSig
+      `mappend` " does not correspond to public key " `mappend` show svfSubmissionSigKey
+
 -- | Validate private block.
--- Block is valid if
--- 1. Header matches body
--- 2. All transactions have valid signatures
--- 3. Transaction signature public key correspond to student public key
-validatePrivateBlk :: PrivateBlock -> Either [Text] ()
+-- Private block is valid iff
+-- 1. Header merkle signature matches computed body merkle signature
+-- 2. All submission in block have valid signatures
+-- 3. All submission signature public keys corresponds to student public key
+--    for given submission
+validatePrivateBlk :: PrivateBlock -> Either [BlockValidationFailure] ()
 validatePrivateBlk pb =
-    let txs = pb ^. pbBody . pbbTxs
-        merkleRoot = getMerkleRoot (fromFoldable txs)
-        headerVer = validateTxHeader (_pbHeader pb) merkleRoot
-        blockValid = verifyGeneric (headerVer <> concatMap validateTx txs)
-    in verResToMonadError toList blockValid
-  where validateTxHeader (PrivateBlockHeader {..}) merkleRoot =
-          [(_pbhBodyProof == merkleRoot, "Header signature missmatch")]
-
-        validateTx (PrivateTx {..}) =
+    let subs =_pbbTxs (_pbBody pb)
+        merkleRoot = getMerkleRoot (fromFoldable subs)
+        headerVer = validateHeader (_pbHeader pb) merkleRoot
+        blockValid = filter ((==False) . fst) (headerVer <> concatMap validateSub subs)
+    in case blockValid of
+         [] -> Right ()
+         x  -> Left  (map snd x)
+  where validateHeader (PrivateBlockHeader {..}) merkleRoot =
+          [(_pbhBodyProof == merkleRoot,
+            MerkleSignatureMismatch { bvfExpectedSig     = _pbhBodyProof
+                                    , bvfActualMerkleSig =  merkleRoot
+                                    })]
+
+        validateSub (PrivateTx {..}) =
           let submission = ssSubmission _ptxSignedSubmission
               witnessKey = _swKey (ssWitness _ptxSignedSubmission)
               witnessSign = _swSig (ssWitness _ptxSignedSubmission)
```

Comment:
The `_swKey` looks like a field from datatype which has lens generated.
If it is, I recommend you using the lenses, not the field direclty, like that:
```haskell
ssWitness^.ptxSignedSubmission.swSig
```
That way you don't have to import things, which names start from underscope, and are able to call `use[s]`, `zoom` and other powerful combinators from `Control.Lens`.

Reply:
Actually, these were not lenses. I took the liberty and made lenses, hope it was ok to do so